src/lib/optimizer/strategy/predicate_split_up_rule.cpp

```diff
@@ -2,44 +2,90 @@
-
+  auto predicate_nodes = std::vector<std::shared_ptr<PredicateNode>>{};
   visit_lqp(root, [&](const auto& sub_node) {
     if (const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(sub_node)) {
-      const auto flat_conjunction = flatten_logical_expressions(predicate_node->predicate(), LogicalOperator::And);
-
-      if (flat_conjunction.size() > 1) {
-        predicate_nodes_to_flat_conjunctions.emplace_back(predicate_node, flat_conjunction);
-      }
+      predicate_nodes.emplace_back(predicate_node);
     }
-
     return LQPVisitation::VisitInputs;
   });
 
+  // _split_conjunction() and _split_disjunction() split up logical expressions by calling each other recursively
+  for (const auto& predicate_node : predicate_nodes) {
+    if (!_split_conjunction(predicate_node)) {
+      // If there is no conjunction at the top level, try to split disjunction first
+      _split_disjunction(predicate_node);
+    }
+  }
+}
+
+bool PredicateSplitUpRule::_split_conjunction(const std::shared_ptr<PredicateNode>& predicate_node) const {
+  const auto flat_conjunction = flatten_logical_expressions(predicate_node->predicate(), LogicalOperator::And);
+  if (flat_conjunction.size() <= 1) {
+    return false;
+  }
+
   /**
-   * Step 2:
-   *    - Split up qualifying PredicateNodes into multiple consecutive PredicateNodes. We have to do this in a
-   *      second pass, because manipulating the LQP within `visit_lqp()`, while theoretically possible, is prone to
-   *      bugs.
+   * Split up PredicateNode with conjunctive chain (e.g., `PredicateNode(a AND b AND c)`) as its scan expression into
+   * multiple consecutive PredicateNodes (e.g. `PredicateNode(c) -> PredicateNode(b) -> PredicateNode(a)`).
    */
-  for (const auto& [predicate_node, flat_conjunction] : predicate_nodes_to_flat_conjunctions) {
-    for (const auto& predicate_expression : flat_conjunction) {
-      lqp_insert_node(predicate_node, LQPInputSide::Left, PredicateNode::make(predicate_expression));
-    }
-    lqp_remove_node(predicate_node);
+  for (const auto& predicate_expression : flat_conjunction) {
+    const auto& new_predicate_node = PredicateNode::make(predicate_expression);
+    lqp_insert_node(predicate_node, LQPInputSide::Left, new_predicate_node);
+    _split_disjunction(new_predicate_node);
+  }
+  lqp_remove_node(predicate_node);
+
+  return true;
+}
+
+void PredicateSplitUpRule::_split_disjunction(const std::shared_ptr<PredicateNode>& predicate_node) const {
+  if (!_should_split_disjunction) {
+    return;
+  }
+
+  const auto flat_disjunction = flatten_logical_expressions(predicate_node->predicate(), LogicalOperator::Or);
+  if (flat_disjunction.size() <= 1) {
+    return;
+  }
+
+  /**
+   * Split up PredicateNode with disjunctive chain (e.g., `PredicateNode(a OR b OR c)`) as their scan expression into
+   * n-1 consecutive UnionNodes and n PredicateNodes.
+   */
+  auto previous_union_node = UnionNode::make(UnionMode::Positions);
```

Comment:
Why `previous` and not `current`?

Reply:
Good point. The variable names became misleading here. I changed them to make their roles in each step clearer.